File: ros2_ws/src/fastumi_usb_camera/src/usb_camera_node.cpp
/**
 * @file usb_camera_node.cpp
 * @brief 通过共享 V4L2 采集核心发布 USB 相机原始图像或原生 JPEG。
 */

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdint>
#include <functional>
#include <memory>
#include <stdexcept>
#include <string>
#include <thread>
#include <utility>
#include <vector>

#include <rcl_interfaces/msg/set_parameters_result.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/compressed_image.hpp>
#include <sensor_msgs/msg/image.hpp>

#include "fastumi_usb_camera/configuration.hpp"
#include "fastumi_usb_camera/image_message.hpp"
#include "fastumi_usb_camera/jpeg_frame.hpp"
#include "fastumi_usb_camera/latest_frame_buffer.hpp"
#include "fastumi_usb_camera/v4l2_camera.hpp"

namespace fastumi_usb_camera
{

using SteadyClock = std::chrono::steady_clock;

/** @brief 保存一帧原生 JPEG 及其主机接收时间。 */
struct CapturedJpeg
{
  builtin_interfaces::msg::Time stamp;  ///< ROS 时钟下的帧接收时间。
  std::vector<uint8_t> data;  ///< 从 V4L2 缓冲区复制出的完整 JPEG 字节。
};

/** @brief 发布 V4L2 USB 相机 raw 或原生 JPEG 图像的 ROS 2 节点。 */
class UsbCameraNode : public rclcpp::Node
{
public:
  /** @brief 声明只读配置并创建与输出模式对应的发布器。 */
  UsbCameraNode()
  : Node("usb_camera_node")
  {
    config_.width = declare_parameter<int64_t>("width", config_.width);
    config_.height = declare_parameter<int64_t>("height", config_.height);
    config_.fps = declare_parameter<int64_t>("fps", config_.fps);
    config_.frame_timeout_seconds = declare_parameter<double>(
      "frame_timeout_seconds", config_.frame_timeout_seconds);
    config_.video_device = declare_parameter<std::string>("video_device", "");
    config_.frame_id = declare_parameter<std::string>("frame_id", config_.frame_id);
    publish_compressed_ = declare_parameter<bool>("publish_compressed", false);
    validate_configuration(config_);

    rclcpp::QoS qos(rclcpp::KeepLast(5));  ///< 与既有接口一致的传感器 QoS。
    qos.best_effort();
    qos.durability_volatile();
    if (publish_compressed_) {
      compressed_publisher_ = create_publisher<sensor_msgs::msg::CompressedImage>(
        "image_raw/compressed", qos);
    } else {
      image_publisher_ = create_publisher<sensor_msgs::msg::Image>("image_raw", qos);
    }
    parameter_callback_ = add_on_set_parameters_callback(
      [](const std::vector<rclcpp::Parameter> &) {
        rcl_interfaces::msg::SetParametersResult result;  ///< 运行时参数拒绝结果。
        result.successful = false;
        result.reason = "USB camera parameters are read-only after startup";
        return result;
      });
    watchdog_timer_ = create_wall_timer(
      std::chrono::milliseconds(500), std::bind(&UsbCameraNode::check_camera, this));
    diagnostics_timer_ = create_wall_timer(
      std::chrono::seconds(5), std::bind(&UsbCameraNode::report_diagnostics, this));
  }

  /** @brief 停止相机并等待发布线程退出。 */
  ~UsbCameraNode() override
  {
    stop();
  }

  /** @brief 启动发布工作线程和共享 V4L2 相机。 */
  void start()
  {
    if (started_.exchange(true)) {
      throw std::logic_error("camera node has already started");
    }
    stopping_ = false;
    last_frame_steady_ns_.store(steady_now_ns());
    worker_ = std::thread(&UsbCameraNode::process_frames, this);
    try {
      camera_ = std::make_unique<V4l2Camera>(
        config_, [this](const uint8_t * data, size_t size) {capture_frame(data, size);});
    } catch (...) {
      frame_buffer_.close();
      worker_.join();
      started_ = false;
      throw;
    }
    RCLCPP_INFO(
      get_logger(), "Streaming V4L2 %s MJPEG %ldx%ld@%ld -> %s",
      camera_->device_path().c_str(), config_.width, config_.height, config_.fps,
      publish_compressed_ ? "image_raw/compressed" : "image_raw");
  }

  /** @brief 幂等停止采集，释放设备并结束发布线程。 */
  void stop() noexcept
  {
    if (started_.exchange(false)) {
      stopping_ = true;
      if (camera_) {
        camera_->stop();
        camera_.reset();
      }
      frame_buffer_.close();
      if (worker_.joinable()) {
        worker_.join();
      }
    }
  }

  /** @return 节点是否遇到需要退出的采集或发布错误。 */
  bool failed() const noexcept
  {
    return fatal_error_.load();
  }

private:
  /** @return 当前单调时钟纳秒数。 */
  static int64_t steady_now_ns()
  {
    return std::chrono::duration_cast<std::chrono::nanoseconds>(
      SteadyClock::now().time_since_epoch()).count();
  }

  /** @brief 在 V4L2 线程复制完整 JPEG 并覆盖旧待处理帧。 */
  void capture_frame(const uint8_t * data, size_t size) noexcept
  {
    if (stopping_ || data == nullptr || size == 0U) {
      callback_invalid_count_.fetch_add(1);
      return;
    }
    try {
      CapturedJpeg captured;  ///< 即将进入最新帧缓存的 JPEG 数据。
      captured.stamp = now();
      captured.data.assign(data, data + size);
      captured_count_.fetch_add(1);
      last_frame_steady_ns_.store(steady_now_ns());
      frame_buffer_.push(std::move(captured));
    } catch (...) {
      set_fatal_error("failed to copy a frame from the V4L2 capture thread");
    }
  }

  /** @brief 在独立线程解码或直接发布最新 JPEG 帧。 */
  void process_frames() noexcept
  {
    try {
      while (auto captured = frame_buffer_.wait_and_take()) {
        if (publish_compressed_) {
          sensor_msgs::msg::CompressedImage message = make_jpeg_message(
            std::move(captured->data), captured->stamp,
            config_.frame_id);  ///< 原生 JPEG ROS 消息。
          compressed_publisher_->publish(std::move(message));
          published_count_.fetch_add(1);
          continue;
        }

        cv::Mat bgr = decode_jpeg_frame(
          captured->data, static_cast<int>(config_.width),
          static_cast<int>(config_.height));  ///< 当前帧的 BGR 解码结果。
        if (bgr.empty()) {
          decode_failure_count_.fetch_add(1);
          continue;
        }
        sensor_msgs::msg::Image message = make_bgr_image_message(
          bgr, captured->stamp, config_.frame_id);  ///< 待发布的 raw 图像消息。
        image_publisher_->publish(std::move(message));
        published_count_.fetch_add(1);
      }
    } catch (const std::exception & error) {
      set_fatal_error(std::string("frame processing failed: ") + error.what());
    } catch (...) {
      set_fatal_error("frame processing failed with an unknown error");
    }
  }

  /** @brief 记录首个致命错误并请求主循环退出。 */
  void set_fatal_error(const std::string & message) noexcept
  {
    if (!fatal_error_.exchange(true)) {
      RCLCPP_FATAL(get_logger(), "%s", message.c_str());
    }
  }

  /** @brief 检查采集线程错误和连续无帧超时。 */
  void check_camera()
  {
    if (!started_ || stopping_) {
      return;
    }
    if (camera_ && camera_->failed()) {
      set_fatal_error("V4L2 camera capture failed: " + camera_->error_message());
      return;
    }
    const int64_t silence_ns =
      steady_now_ns() - last_frame_steady_ns_.load();  ///< 距离最后有效帧的时间。
    if (silence_ns > static_cast<int64_t>(config_.frame_timeout_seconds * 1.0e9)) {
      set_fatal_error(
        "no valid V4L2 camera frame received for " +
        std::to_string(config_.frame_timeout_seconds) + " seconds");
    }
  }

  /** @brief 周期记录采集、发布、覆盖、坏帧和解码失败数量。 */
  void report_diagnostics()
  {
    const uint64_t camera_invalid = camera_ ? camera_->invalid_count() : 0U;  ///< 驱动坏帧数。
    RCLCPP_INFO(
      get_logger(),
      "frames captured=%lu published=%lu overwritten=%lu invalid=%lu decode_failed=%lu",
      captured_count_.load(), published_count_.load(),
      frame_buffer_.overwritten_count(), camera_invalid + callback_invalid_count_.load(),
      decode_failure_count_.load());
  }

  CameraConfiguration config_;  ///< 设备选择、采集模式及消息坐标系配置。
  bool publish_compressed_{false};  ///< true 时直接发布相机原生 JPEG。
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_publisher_;  ///< raw 发布器。
  rclcpp::Publisher<sensor_msgs::msg::CompressedImage>::SharedPtr
    compressed_publisher_;  ///< 原生 JPEG 发布器。
  rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr
    parameter_callback_;  ///< 拒绝运行时参数修改的回调句柄。
  LatestFrameBuffer<CapturedJpeg> frame_buffer_;  ///< 只保存最新待处理帧的缓存。
  std::unique_ptr<V4l2Camera> camera_;  ///< 唯一 V4L2 设备资源所有者。
  std::thread worker_;  ///< 图像解码和发布线程。
  rclcpp::TimerBase::SharedPtr watchdog_timer_;  ///< 采集异常和无帧超时检查定时器。
  rclcpp::TimerBase::SharedPtr diagnostics_timer_;  ///< 帧统计输出定时器。
  std::atomic<bool> started_{false};  ///< 节点是否已启动相机。
  std::atomic<bool> stopping_{false};  ///< 节点是否正在停止并拒绝新帧。
  std::atomic<bool> fatal_error_{false};  ///< 是否发生需要结束进程的错误。
  std::atomic<int64_t> last_frame_steady_ns_{0};  ///< 最后有效帧的单调时间。
  std::atomic<uint64_t> captured_count_{0};  ///< 已从 V4L2 接收的完整 JPEG 帧数量。
  std::atomic<uint64_t> published_count_{0};  ///< 已发布图像数量。
  std::atomic<uint64_t> callback_invalid_count_{0};  ///< 回调层拒绝的空帧数量。
  std::atomic<uint64_t> decode_failure_count_{0};  ///< JPEG 解码或尺寸校验失败数量。
};

}  // namespace fastumi_usb_camera

namespace
{

volatile std::sig_atomic_t stop_requested = 0;  ///< SIGINT/SIGTERM 请求的退出标志。

/** @brief 将进程终止信号转换为主循环可轮询的标志。 */
void request_stop(int signal_number)
{
  (void)signal_number;
  stop_requested = 1;
}

}  // namespace

/** @brief 创建 USB 相机节点并在退出前确定性释放 V4L2 设备。 */
int main(int argc, char ** argv)
{
  rclcpp::init(
    argc, argv, rclcpp::InitOptions(), rclcpp::SignalHandlerOptions::None);
  std::signal(SIGINT, request_stop);
  std::signal(SIGTERM, request_stop);
  try {
    bool failed = false;  ///< 节点退出码是否应表示运行失败。
    {
      auto node = std::make_shared<fastumi_usb_camera::UsbCameraNode>();  ///< 相机节点实例。
      node->start();
      rclcpp::executors::SingleThreadedExecutor executor;  ///< 定时器与参数回调执行器。
      executor.add_node(node);
      while (!stop_requested && !node->failed()) {
        executor.spin_once(std::chrono::milliseconds(100));
      }
      failed = node->failed();
      executor.remove_node(node);
      node->stop();
      node.reset();
    }
    rclcpp::shutdown();
    return failed ? 1 : 0;
  } catch (const std::exception & error) {
    RCLCPP_FATAL(rclcpp::get_logger("usb_camera_node"), "%s", error.what());
    rclcpp::shutdown();
    return 1;
  }
}

File: ros2_ws/src/fastumi_usb_camera/fastumi_usb_camera/usb_camera_node.py
"""按启动模式将 UVC MJPEG 帧发布为 ROS 2 原始或压缩图像。"""

from __future__ import annotations

import signal
import threading
import time
from typing import Callable

import cv2
from cv_bridge import CvBridge
import numpy as np
from rcl_interfaces.msg import SetParametersResult
import rclpy
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, HistoryPolicy, QoSProfile, ReliabilityPolicy
from sensor_msgs.msg import CompressedImage, Image

from fastumi_usb_camera.capture import UvcCamera
from fastumi_usb_camera.latest_frame import CapturedFrame, LatestFrame


# 连续无完整帧的设备错误门限，单位为秒。
FRAME_TIMEOUT_SECONDS = 5.0
# 解码失败告警的最小间隔，单位为秒。
WARNING_INTERVAL_SECONDS = 5.0


class UsbCameraNode(Node):
    """在 ROS 线程按模式发布图像，在 UVC 线程仅缓存最新帧。"""

    def __init__(self, camera_factory: Callable[..., UvcCamera] = UvcCamera) -> None:
        """声明启动参数、创建发布器并打开指定 USB 相机。"""
        super().__init__("usb_camera_node")
        self.declare_parameter("vendor_id", 0x1BCF)
        self.declare_parameter("product_id", 0x28C4)
        self.declare_parameter("device_uid", "")
        self.declare_parameter("video_device", "")
        self.declare_parameter("width", 1280)
        self.declare_parameter("height", 960)
        self.declare_parameter("fps", 30)
        self.declare_parameter("frame_id", "usb_camera_optical_frame")
        self.declare_parameter("publish_compressed", False)
        self.declare_parameter("publish_reliability", "reliable")
        self.add_on_set_parameters_callback(self._reject_runtime_parameters)

        # 这些参数只在创建节点时读取，不支持运行时改变设备或发布器。
        camera_config = {
            name: self._positive_int(name)
            for name in ("width", "height", "fps")
        }
        device_uid = self.get_parameter("device_uid").value
        if not isinstance(device_uid, str):
            raise ValueError("device_uid 必须是字符串")
        camera_config["device_uid"] = device_uid.strip()
        video_device = self.get_parameter("video_device").value
        if not isinstance(video_device, str):
            raise ValueError("video_device 必须是字符串")
        camera_config["video_device"] = video_device.strip()
        if not camera_config["video_device"]:
            camera_config["vendor_id"] = self._positive_int("vendor_id")
            camera_config["product_id"] = self._positive_int("product_id")
        self._width = camera_config["width"]
        self._height = camera_config["height"]
        self._frame_id = str(self.get_parameter("frame_id").value).strip()
        if not self._frame_id:
            raise ValueError("frame_id 不能为空")
        self._publish_compressed = bool(
            self.get_parameter("publish_compressed").value
        )
        publish_reliability = str(self.get_parameter("publish_reliability").value).strip().lower()
        if publish_reliability not in ("reliable", "best_effort"):
            raise ValueError("publish_reliability 只能是 reliable 或 best_effort")
        qos = QoSProfile(
            history=HistoryPolicy.KEEP_LAST,
            depth=20,
            reliability=(ReliabilityPolicy.RELIABLE if publish_reliability == "reliable"
                         else ReliabilityPolicy.BEST_EFFORT),
            durability=DurabilityPolicy.VOLATILE,
        )
        self._image_publisher = (
            None if self._publish_compressed
            else self.create_publisher(Image, "image_raw", qos)
        )
        self._compressed_publisher = (
            self.create_publisher(CompressedImage, "image_raw/compressed", qos)
            if self._publish_compressed else None
        )
        self._bridge = None if self._publish_compressed else CvBridge()
        self._frames = LatestFrame()
        self.published_count = 0
        self.decode_failed_count = 0
        self._last_warning = 0.0
        self._last_complete_at = time.monotonic()
        self._started = False
        self._closed = False
        self._guard = self.create_guard_condition(self._publish_latest)
        self._watchdog = self.create_timer(0.5, self._check_camera)
        self._diagnostics = self.create_timer(5.0, self._report_counts)
        try:
            self._camera = camera_factory(**camera_config)
        except Exception:
            self.destroy_node()
            raise
        camera_label = camera_config["video_device"] or (
            f"{camera_config['vendor_id']:04x}:{camera_config['product_id']:04x}"
        )
        output_topic = (
            "image_raw/compressed" if self._publish_compressed else "image_raw"
        )
        self.get_logger().info(
            f"USB 相机 {camera_label}，{self._width}x{self._height}@"
            f"{camera_config['fps']} FPS，发布 {output_topic}"
        )

    def _positive_int(self, name: str) -> int:
        """获取并校验正整数相机参数。"""
        value = self.get_parameter(name).value
        if type(value) is not int or value <= 0:
            raise ValueError(f"{name} 必须是正整数")
        if name in ("vendor_id", "product_id") and value > 0xFFFF:
            raise ValueError(f"{name} 必须在 USB 16 位标识范围内")
        return value

    def _reject_runtime_parameters(self, _parameters: list) -> SetParametersResult:
        """拒绝运行时修改设备和发布接口参数。"""
        return SetParametersResult(
            successful=False, reason="USB 相机参数仅支持在节点启动时配置"
        )

    def start(self) -> None:
        """在相机线程开始接收 JPEG 帧。"""
        if self._started or self._closed:
            raise RuntimeError("相机节点不能重复启动")
        self._last_complete_at = time.monotonic()
        self._started = True
        self._camera.start(self._receive_frame)

    def _receive_frame(self, jpeg: bytes) -> None:
        """在采集线程记录主机 ROS 时间戳并覆盖旧待处理帧。"""
        stamp = self.get_clock().now().to_msg()
        self._last_complete_at = time.monotonic()
        self._frames.put(CapturedFrame(stamp, jpeg))
        self._guard.trigger()

    def _publish_latest(self) -> None:
        """在 ROS 线程发布最新帧；压缩模式直接转发 JPEG 字节。"""
        frame = self._frames.take()
        if frame is None:
            return
        if self._compressed_publisher is not None:
            compressed = CompressedImage()
            compressed.header.stamp = frame.stamp
            compressed.header.frame_id = self._frame_id
            compressed.format = "jpeg"
            compressed.data = frame.jpeg
            self._compressed_publisher.publish(compressed)
            self.published_count += 1
            return
        image = cv2.imdecode(np.frombuffer(frame.jpeg, dtype=np.uint8),
                             cv2.IMREAD_COLOR)
        if image is None or image.shape[:2] != (self._height, self._width):
            self.decode_failed_count += 1
            now = time.monotonic()
            if now - self._last_warning >= WARNING_INTERVAL_SECONDS:
                self.get_logger().warning("JPEG 解码失败或图像尺寸与采集模式不符，已跳过")
                self._last_warning = now
            return
        raw = self._bridge.cv2_to_imgmsg(image, encoding="bgr8")
        raw.header.stamp = frame.stamp
        raw.header.frame_id = self._frame_id
        self._image_publisher.publish(raw)
        self.published_count += 1

    def _check_camera(self) -> None:
        """发现采集异常、意外停止或连续无帧时终止节点。"""
        if not self._started or self._closed:
            return
        if self._camera.last_error is not None:
            raise RuntimeError("UVC 相机采集失败") from self._camera.last_error
        if self._camera.stopped():
            raise RuntimeError("UVC 相机采集意外停止")
        if time.monotonic() - self._last_complete_at >= FRAME_TIMEOUT_SECONDS:
            raise RuntimeError("UVC 相机连续 5 秒没有完整图像帧")

    def _report_counts(self) -> None:
        """周期记录采集、发布、积压覆盖和坏帧数量。"""
        self.get_logger().info(
            f"帧统计: 采集={self._frames.captured_count}, "
            f"发布={self.published_count}, 覆盖={self._frames.overwritten_count}, "
            f"不完整={self._camera.incomplete_count}, "
            f"解码失败={self.decode_failed_count}"
        )

    def close(self) -> None:
        """先停止并释放相机，再由调用方销毁 ROS 节点。"""
        if self._closed:
            return
        self._closed = True
        self._camera.close()


def main(args: list[str] | None = None) -> None:
    """启动相机节点，运行 ROS 事件循环并在退出时释放设备。"""
    rclpy.init(args=args)
    node: UsbCameraNode | None = None
    stop_event = threading.Event()

    def request_stop(_signum: int, _frame: object) -> None:
        """将 SIGINT/SIGTERM 转换为事件，避免清理时二次中断。"""
        stop_event.set()

    previous_sigint = signal.signal(signal.SIGINT, request_stop)
    previous_sigterm = signal.signal(signal.SIGTERM, request_stop)
    try:
        node = UsbCameraNode()
        node.start()
        while rclpy.ok() and not stop_event.is_set():
            rclpy.spin_once(node, timeout_sec=0.1)
    finally:
        try:
            if node is not None:
                try:
                    node.close()
                finally:
                    node.destroy_node()
        finally:
            if rclpy.ok():
                rclpy.shutdown()
            signal.signal(signal.SIGINT, previous_sigint)
            signal.signal(signal.SIGTERM, previous_sigterm)


if __name__ == "__main__":
    main()
